tephra-core/src/test/java/co/cask/tephra/TransactionManagerTest.java

```diff
@@ -138,4 +138,45 @@ public void testTransactionCleanup() throws Exception {
       txm.stopAndWait();
     }
   }
+
+  @Test
+  public void testLongTransactionCleanup() throws Exception {
+    conf.setInt(TxConstants.Manager.CFG_TX_CLEANUP_INTERVAL, 3);
+    conf.setInt(TxConstants.Manager.CFG_TX_LONG_TIMEOUT, 2);
+    // using a new tx manager that cleans up
+    TransactionManager txm = new TransactionManager
+      (conf, new InMemoryTransactionStateStorage(), new TxMetricsCollector());
+    txm.startAndWait();
+    try {
+      Assert.assertEquals(0, txm.getInvalidSize());
+      Assert.assertEquals(0, txm.getCommittedSize());
+      
+      // start a long running transaction
+      Transaction tx1 = txm.startLong();
+      
+      Assert.assertEquals(0, txm.getInvalidSize());
+      Assert.assertEquals(0, txm.getCommittedSize());
+
+      // sleep longer than the cleanup interval
+      TimeUnit.SECONDS.sleep(5);
+
+      // transaction should now be invalid
+      Assert.assertEquals(1, txm.getInvalidSize());
+      Assert.assertEquals(0, txm.getCommittedSize());
+
+      // cannot commit transaction as it was timed out
+      try {
+        txm.canCommit(tx1, Collections.singleton(new byte[] { 0x11 }));
+        Assert.fail();
+      } catch (TransactionNotInProgressException e) {
+        // expected
+      }
+      
+      txm.abort(tx1);
```

Comment:
This test made me look at the TransactionManager.doAbort() handling, and it looks like we'll have a change in behavior there due to this patch.

Previously, long running transactions would always be in progress since they weren't timed out, so we would always hit the first branch in that method -- resulting in the long running transaction going into the invalid list and staying there.  Now, with this change, if the long running transaction is timed out, and still calls abort(), it would hit the second branch in the conditional logic, resulting in the transaction ID being removed from the invalid list.  If we are making the assumption that the long running transaction's edits cannot be rolled back, then this is not what we want.  I think we should preserve the current behavior if we are relying on that in CDAP.